planner/core/logical_plans.go

```diff
@@ -320,6 +322,9 @@ type LogicalSelection struct {
 	// but after we converted to CNF(Conjunctive normal form), it can be
 	// split into a list of AND conditions.
 	Conditions []expression.Expression
+	// HasBeenPushed means whether the selection is applied rule `PushSelDown`,
+	// it's used in cascades planner.
+	HasBeenPushed bool
```

Comment:
We need a general solution to solve the recursive rule applying. There's also other cases that would cause endless loop.
You can add a TODO here first.

Reply:
OK.

Reply:
This will be done by `AppliedRuleSet`. When we apply the rule `PushSelDownJoin`, we should add this rule to `AppliedRuleSet` of the new created Selection, so that the new Selection will not match this rule again.

Reply:
Where is the `AppliedRuleSet`?

Reply:
I'm still working on it >.<

Reply:
OK...Don't forget to remove this when you finish it.  0.0